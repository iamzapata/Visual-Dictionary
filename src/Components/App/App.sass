@charset "utf-8"
@import "~bulma/bulma"
@import "~slick-carousel/slick/slick.css"
@import "~slick-carousel/slick/slick-theme.css"

=font-face($family, $path, $svg, $weight: normal, $style: normal)
  @font-face
    font-family: $family
    src: url('#{$path}.eot')
    src: url('#{$path}.eot?#iefix') format('embedded-opentype'), url('#{$path}.woff') format('woff'), url('#{$path}.ttf') format('truetype'), url('#{$path}.svg##{$svg}') format('svg')
    font-weight: $weight
    font-style: $style

/* nunito-regular - latin */
+font-face('Nunito', '../../fonts/nunito-v9-latin-regular', 'Nunito', 400, 'normal')

/* nunito-600 - latin */
+font-face('Nunito', '../../fonts/nunito-v9-latin-600', 'Nunito', 600, 'normal')

/* nunito-700 - latin */
+font-face('Nunito', '../../fonts/nunito-v9-latin-600', 'Nunito', 600, 'normal')


html
  font-family: "Nunito", sans-serif
  min-height: 100%
  display: flex
  align-items: flex-start
  justify-content: center

body
  font-family: "Nunito", sans-serif

.App
  position: relative
  min-width: 490px
  max-width: 490px
  margin: 100px 0 300px 0

  .LoadingSpinner
    position: absolute
    display: flex
    justify-content: center
    width: 100%
    top: 120px
    z-index: 10000
