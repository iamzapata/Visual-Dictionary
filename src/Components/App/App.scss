@charset "utf-8";
@import "~bulma/bulma";
@import "~slick-carousel/slick/slick.css";
@import "~slick-carousel/slick/slick-theme.css";
@import "src/scss/fontFace";

/* nunito-regular - latin */
@include font-face(
  "Nunito",
  "../../fonts/nunito-v9-latin-regular",
  "Nunito",
  400,
  "normal"
);

/* nunito-600 - latin */
@include font-face(
  "Nunito",
  "../../fonts/nunito-v9-latin-600",
  "Nunito",
  600,
  "normal"
);

/* nunito-700 - latin */
@include font-face(
  "Nunito",
  "../../fonts/nunito-v9-latin-600",
  "Nunito",
  600,
  "normal"
);

html {
  font-family: "Nunito", sans-serif;
  min-height: 100%;
  display: flex;
  align-items: flex-start;
  justify-content: center;

  body {
    font-family: "Nunito", sans-serif;
  }
}

.App {
  position: relative;
  min-width: 490px;
  max-width: 490px;
  margin: 100px 0 300px 0;
}

.LoadingSpinner {
  position: absolute;
  display: flex;
  justify-content: center;
  width: 100%;
  top: 120px;
  z-index: 10000;
}
