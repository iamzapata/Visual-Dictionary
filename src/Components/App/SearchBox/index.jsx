import React, { Component } from "react"
import { func } from "prop-types"

class SearchBox extends Component {
  state = {
    inputValue: "",
    showEmptySearchError: false
  }

  onSubmitSearch = ev => {
    ev.preventDefault()
    const { inputValue } = this.state
    if (!inputValue.length) {
      this.setState({ showEmptySearchError: true })
      return
    }
    this.props.searchWord(inputValue)
  }

  onInputChange = ev => {
    ev.stopPropagation()
    const { value: inputValue } = ev.target
    if(inputValue.length) {
      this.setState({inputValue, showEmptySearchError: false})
      return
    }
    //const showEmptySearchError = inputValue.length ? false : true
    this.setState({ inputValue})
  }

  render() {
    const { inputValue, showEmptySearchError } = this.state
    return (
      <div className="SearchBox">
        <form
          className="SearchBox__Form"
          onSubmit={ev => this.onSubmitSearch(ev)}
        >
          <input
            type="text"
            className="SearchBox__Input"
            onChange={this.onInputChange}
            placeholder="Define ..."
            value={inputValue}
          />
          {showEmptySearchError && (
            <span className="ErrorMessage">
              Please type the word you want to search for
            </span>
          )}
          <button>Define</button>
        </form>
      </div>
    )
  }
}

SearchBox.propTypes = {
  searchWord: func.isRequired
}

export default SearchBox
