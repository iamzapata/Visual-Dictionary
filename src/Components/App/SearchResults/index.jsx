import React from "react"

const SearchResults = () => <div className="SearchResults" />

export default SearchResults
