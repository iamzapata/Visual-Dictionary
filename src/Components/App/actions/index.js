import ActionTypes from "Components/App/actionTypes/index"
import request from "utils/request/index"

export const searchWordRequest = searchQuery => ({
  type: ActionTypes.SEARCH_WORD_REQUEST,
  err: null,
  isLoading: true,
  searchQuery
})

export function searchWordSuccess(response, imageResults) {
  const { results } = response
  return {
    type: ActionTypes.SEARCH_WORD_SUCCESS,
    err: null,
    isLoading: false,
    results,
    imageResults
  }
}

export function searchWordFailure(err) {
  return {
    type: ActionTypes.SEARCH_WORD_FAILURE,
    err,
    isLoading: false
  }
}

function displaySuggestions(suggestions) {
  return {
    type: ActionTypes.SEARCH_WORD_MISSPELL_SUGGESTIONS,
    suggestions
  }
}

function searchImages(queryString) {
  return fetch(
    `https://www.googleapis.com/customsearch/v1?
    key=${GOOGLE_SEARCH_KEY}&cx=${GOOGLE_SEARCH_ENGINE}&searchType=image&q=${queryString}
    `
  ).then(response => (response.items ? response.items : []))
}

export default function searchWord(queryString) {
  return async dispatch => {
    dispatch(searchWordRequest(queryString))
    try {
      const entries = await request(`define/${queryString}`)
      const relatedImages = await searchImages(queryString)
      dispatch(searchWordSuccess(entries, relatedImages))
    } catch (err) {
      if (err.message === "404") {
        const suggestions = await request(`suggestions/${queryString}`
        ).then(response => response.results)
        dispatch(displaySuggestions(suggestions))
      }
      dispatch(searchWordFailure(err))
    }
  }
}
