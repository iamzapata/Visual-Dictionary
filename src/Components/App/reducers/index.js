import { combineReducers } from "redux"
import SearchStore from "./SearchStore"

const rootReducer = combineReducers({
  SearchStore
})

export default rootReducer
