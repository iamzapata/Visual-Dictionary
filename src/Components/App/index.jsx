import React from "react"
import { func, oneOfType, node, string, shape, bool, arrayOf } from "prop-types"
import { FadeLoader } from "react-spinners"
import SearchBox from "Components/App/SearchBox"
import SearchResults from "Components/App/SearchResults"
import { connect } from "react-redux"
import searchWord from "Components/App/actions"
import {
  lexicalEntriesSelector,
  imageResultsSelector
} from "Components/App/reducers/selectors"
import "./App.sass"

export const App = ({
  header,
  isLoading,
  searchWord,
  SearchStore,
  imageResults,
  lexicalEntries
}) => (
  <div className="App">
    <h2 className="title is-2 has-text-centered">{header}</h2>
    <SearchBox searchWord={searchWord} SearchStore={SearchStore} />
    {isLoading && (
      <div className="LoadingSpinner">
        <FadeLoader />
      </div>
    )}
    <SearchResults results={lexicalEntries} imageResults={imageResults} />
  </div>
)

App.propTypes = {
  searchWord: func.isRequired,
  header: oneOfType([node, string]).isRequired,
  SearchStore: shape({
    isLoading: bool,
    results: arrayOf(shape({ shape }))
  }).isRequired,
  lexicalEntries: arrayOf(shape({})).isRequired,
  imageResults: arrayOf(shape({})).isRequired,
  isLoading: bool.isRequired
}

const mapDispatchToProps = {
  searchWord
}

const mapStateToProps = state => ({
  SearchStore: state.SearchStore,
  lexicalEntries: lexicalEntriesSelector(state),
  isLoading: state.SearchStore.isLoading,
  imageResults: imageResultsSelector(state)
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)
