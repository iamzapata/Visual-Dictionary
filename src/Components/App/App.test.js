import React from "react";
import { render } from "react-testing-library";
import App from "./index.jsx";

describe("<App />", () => {
  const store = {
    subscribe: () => null,
    dispatch: () => null,
    getState: () => null
  };
  let props = { header: "Visual Dictionary", store };

  it("should render Self: <App />", () => {
    const { container, queryByText } = render(<App {...props} />);
    expect(container.firstChild).toMatchSnapshot();
    const header = queryByText(props.header);
    expect(header.innerHTML).toBe(props.header);
  });

  it("should render a <SearchBox /> component", () => {
    const { container } = render(<App {...props} />);
    const searchBox = container.querySelectorAll(".SearchBox");
    expect(searchBox.length).toBe(1);
  });

  it("should render a <SearchResults /> component", () => {
    const { container } = render(<App {...props} />);
    const results = container.querySelectorAll(".SearchResults");
    expect(results.length).toBe(1);
  });
});
