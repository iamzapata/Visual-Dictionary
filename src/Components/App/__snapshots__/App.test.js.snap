// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Should render self 1`] = `
<div
  class="App"
>
  <h2
    class="title is-2 has-text-centered"
  >
    Visual Dictionary
  </h2>
  <div
    class="SearchBox"
  >
    <form
      class="SearchBox__Form"
    >
      <div
        class="field has-addons"
      >
        <div
          class="SearchBox__Input-Control control"
        >
          <input
            class="SearchBox__Input input"
            placeholder="Define ..."
            type="text"
            value=""
          />
        </div>
        <div
          class="SearchBox__Button-Control control"
        >
          <button
            class="button"
          >
            Search
          </button>
        </div>
      </div>
    </form>
  </div>
  <div
    class="SearchResults"
  >
    <div
      class="SearchResults__Images"
    >
      <div
        class="slick-slider slick-initialized"
      >
        
        <div
          class="slick-list"
        >
          <div
            class="slick-track"
            style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
          />
        </div>
        
        
      </div>
    </div>
    <ul
      class="SearchResults__Definitions"
    />
  </div>
</div>
`;
